### PR TITLE
Optimized Llama 3.x perf with sharded residual

### PR DESCRIPTION
**Pull request by yieldthought**
### Ticket
[14273](https://github.com/tenstorrent/tt-metal/pull/14273)

### Problem description
The shared llama3 codebase used interleaved L1 for the residual during decode. This had lower performance than previous specialized models.

### What's changed
* Sharded the residual path for all combinations of models and devices
* Resolved the L1 corruption issues without deallocate workarounds
* Improved performance over all previous specialised models: 8b n150 is now >24t/s/u and 70b t3k is >15 t/s/u

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/12007682870)
- [x] [Single card demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/11975122584)
- [x] [Single card model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11975114297)
- [ ] [T3K demo + other tests](https://github.com/tenstorrent/tt-metal/actions/runs/11975098695) ([perf rerun](https://github.com/tenstorrent/tt-metal/actions/runs/12009863881))

### CODE REVIEW

**Comment by davorchap** [comment]
Nice !

**Review by cglagovichTT — APPROVED** [review]

**Comment by yieldthought** [comment]
`models/demos/llama3/tests/multimodal/test_llama_cross_attention_transformer_text.py` fails with PCC errors. This passes on main. Investigating.

File: models/demos/llama3/tt/multimodal/llama_cross_attention_transformer_text.py [hunk 1]
```diff
@@ -288,8 +283,9 @@ def forward(
 
         h = self.norm(h, mode=mode)
 
-        if mode == "decode":  # h is expected to be interleaved for the lm head
-            h = ttnn.sharded_to_interleaved(h)
+        # TODO: Switch to using dram-sharded LM haed and remove this
```
Comment (mtairum) [inline models/demos/llama3/tt/multimodal/llama_cross_attention_transformer_text.py:286]: `head` instead of `haed`

**Review by mtairum — APPROVED** [review]

**Comment by yieldthought** [comment]
Note that the async dispatch revert in commit `30364957fe888854bda3ae43a00b1867f7c2291c` issue #15324 decreases the performance by around 10% so we actually see ~22.8 t/s/u 8b N150 for example. Revert that to see the promised perf gains in the description.

**Pull request merged**